store/sqlstore/post_store.go

```diff
@@ -2565,6 +2580,103 @@ func (s *SqlPostStore) SearchPostsForUser(paramsList []*model.SearchParams, user
 	return model.MakePostSearchResults(posts, nil), nil
 }
 
+const lastSearchesLimit = 5
+
+func (s *SqlPostStore) LogRecentSearch(userID string, searchQuery []byte, createAt int64) error {
+	transaction, err := s.GetMasterX().Beginx()
+	if err != nil {
+		return errors.Wrap(err, "begin_transaction")
+	}
+
+	defer finalizeTransactionX(transaction)
+
+	var lastSearchPointer int
+	var queryStr string
+	// get search_pointer
+	// We coalesce to -1 because we want to start from 0
+	if s.DriverName() == model.DatabaseDriverPostgres {
+		queryStr = `SELECT COALESCE((props->>'last_search_pointer')::integer, -1)
+			FROM Users
+			WHERE Id=?`
+	} else {
+		queryStr = `SELECT COALESCE(CAST(JSON_EXTRACT(Props, '$.last_search_pointer') as unsigned), -1)
+			FROM Users
+			WHERE Id=?`
+	}
+	err = transaction.Get(&lastSearchPointer, queryStr, userID)
+	if err != nil {
+		return errors.Wrapf(err, "failed to find last_search_pointer for user=%s", userID)
+	}
+
+	// (ptr+1)%lastSearchesLimit
+	lastSearchPointer = (lastSearchPointer + 1) % lastSearchesLimit
+
+	if s.IsBinaryParamEnabled() {
+		searchQuery = s.AppendBinaryFlag(searchQuery)
+	}
+
+	// insert at pointer
+	query := s.getQueryBuilder().
+		Insert("RecentSearches").
+		Columns("UserId", "SearchPointer", "Query", "CreateAt").
+		Values(userID, lastSearchPointer, searchQuery, createAt)
+
+	if s.DriverName() == model.DatabaseDriverPostgres {
+		query = query.SuffixExpr(sq.Expr("ON CONFLICT (userid, searchpointer) DO UPDATE SET Query = ?, CreateAt = ?", searchQuery, createAt))
+	} else {
+		query = query.SuffixExpr(sq.Expr("ON DUPLICATE KEY UPDATE Query = ?, CreateAt = ?", searchQuery, createAt))
+	}
+
+	queryString, args, err := query.ToSql()
+	if err != nil {
+		return errors.Wrap(err, "log_recent_search_tosql")
+	}
+
+	if _, err2 := transaction.Exec(queryString, args...); err2 != nil {
+		return errors.Wrapf(err2, "failed to upsert recent_search for user=%s", userID)
+	}
+
+	// write ptr on users prop
+	if s.DriverName() == model.DatabaseDriverPostgres {
+		_, err = transaction.Exec(`UPDATE Users
+			SET Props = jsonb_set(Props, $1, $2)
+			WHERE Id = $3`, jsonKeyPath("last_search_pointer"), jsonStringVal(strconv.Itoa(lastSearchPointer)), userID)
+	} else {
+		_, err = transaction.Exec(`UPDATE Users
+			SET Props = JSON_SET(Props, ?, ?)
+			WHERE Id = ?`, "$.last_search_pointer", strconv.Itoa(lastSearchPointer), userID)
+	}
+	if err != nil {
+		return errors.Wrapf(err, "failed to update last_search_pointer for user=%s", userID)
+	}
+
+	if err2 := transaction.Commit(); err2 != nil {
+		return errors.Wrap(err2, "commit_transaction")
+	}
+
+	return nil
+}
+
+func (s *SqlPostStore) GetRecentSearchesForUser(userID string) ([]*model.SearchParams, error) {
+	params := [][]byte{}
+	err := s.GetReplicaX().Select(&params, `SELECT query
+		FROM RecentSearches
+		WHERE UserId=?
+		ORDER BY CreateAt DESC`, userID)
```

Comment:
Looks like we are performing a query on the database without any upper limits. This might be performance intensive if a user creates multiple searches. Can we have a limit on max number of rows?